### PR TITLE
feat(hatchery): manage string jobID and ascode worker model

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18976.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18977.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18981.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18985.1 ■   
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ■

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18985.1 ■   
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ■

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18989.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18991.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test UI ✔

**Review by yesnault — APPROVED** [review]

File: engine/hatchery/vsphere/hatchery.go [hunk 1]
```diff
@@ -133,20 +134,22 @@ func (h *HatcheryVSphere) CheckConfiguration(cfg interface{}) error {
...
 			r.Type == sdk.MemoryRequirement ||
 			r.Type == sdk.HostnameRequirement ||
-			model.ModelVirtualMachine.Cmd == "" {
+			model.GetCmd() == "" {
 			return false
 		}
 	}
 
-	if jobID <= 0 {
+	if jobID == "0" {
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:152]: Convert to int then compare <= 0 ?

File: engine/hatchery/swarm/swarm.go [hunk 2]
```diff
@@ -272,14 +272,14 @@ func (h *HatcherySwarm) SpawnWorker(ctx context.Context, spawnArgs hatchery.Spaw
...
-		memory = spawnArgs.Model.ModelDocker.Memory
+	if spawnArgs.Model.GetDockerMemory() != 0 {
+		memory = spawnArgs.Model.GetDockerMemory()
 	}
 
 	var network, networkAlias string
 	services := []string{}
 
-	if spawnArgs.JobID > 0 {
+	if spawnArgs.JobID != "0" {
```
Comment (richardlt) [inline engine/hatchery/swarm/swarm.go:282]: Same

File: engine/hatchery/swarm/swarm.go [hunk 3]
```diff
@@ -212,7 +212,7 @@ func (h *HatcherySwarm) SpawnWorker(ctx context.Context, spawnArgs hatchery.Spaw
 	ctx, end := telemetry.Span(ctx, "swarm.SpawnWorker")
 	defer end()
 
-	if spawnArgs.JobID == 0 && !spawnArgs.RegisterOnly {
+	if spawnArgs.JobID == "0" && !spawnArgs.RegisterOnly {
```
Comment (richardlt) [inline engine/hatchery/swarm/swarm.go:215]: Same

File: engine/hatchery/local/worker_spawn.go [hunk 4]
```diff
@@ -28,7 +28,7 @@ func (localWorkerRunner) NewCmd(ctx context.Context, command string, args ...str
 func (h *HatcheryLocal) SpawnWorker(ctx context.Context, spawnArgs hatchery.SpawnArguments) error {
 	log.Debug(ctx, "HatcheryLocal.SpawnWorker> %s want to spawn a worker named %s (jobID = %d)", spawnArgs.HatcheryName, spawnArgs.WorkerName, spawnArgs.JobID)
 
-	if spawnArgs.JobID == 0 && !spawnArgs.RegisterOnly {
+	if spawnArgs.JobID == "0" && !spawnArgs.RegisterOnly {
```
Comment (richardlt) [inline engine/hatchery/local/worker_spawn.go:31]: Same

File: engine/hatchery/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -220,13 +222,13 @@ func (h *HatcheryKubernetes) SpawnWorker(ctx context.Context, spawnArgs hatchery
...
 
-	if spawnArgs.JobID == 0 && !spawnArgs.RegisterOnly {
+	if spawnArgs.JobID == "0" && !spawnArgs.RegisterOnly {
 		return sdk.WithStack(fmt.Errorf("no job ID and no register"))
 	}
 
 	var logJob string
-	if spawnArgs.JobID > 0 {
-		logJob = fmt.Sprintf("for workflow job %d,", spawnArgs.JobID)
+	if spawnArgs.JobID != "0" {
```
Comment (richardlt) [inline engine/hatchery/kubernetes/kubernetes.go:230]: Same

File: engine/hatchery/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -220,13 +222,13 @@ func (h *HatcheryKubernetes) SpawnWorker(ctx context.Context, spawnArgs hatchery
 		telemetry.Tag(telemetry.TagWorker, spawnArgs.WorkerName))
 	defer end()
 
-	if spawnArgs.JobID == 0 && !spawnArgs.RegisterOnly {
+	if spawnArgs.JobID == "0" && !spawnArgs.RegisterOnly {
```
Comment (richardlt) [inline engine/hatchery/kubernetes/kubernetes.go:225]: Same

**Review by richardlt — CHANGES_REQUESTED** [review]

File: engine/worker/internal/start.go [hunk 7]
```diff
@@ -12,13 +13,13 @@ import (
...
 
-func StartWorker(ctx context.Context, w *CurrentWorker, bookedJobID int64) (mainError error) {
+func StartWorker(ctx context.Context, w *CurrentWorker, bookedJobID string) (mainError error) {
 	ctx = context.WithValue(ctx, log.Field("permJobID"), bookedJobID)
 
 	log.Info(ctx, "Starting worker %s on job %d", w.Name(), bookedJobID)
 
-	if bookedJobID == 0 {
-		return errors.Errorf("startWorker: bookedJobID is mandatory. val:%d", bookedJobID)
+	if bookedJobID == "0" {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/worker/internal/start.go:21]: done

File: sdk/hatchery/hatchery.go [hunk 8]
```diff
@@ -641,12 +755,12 @@ func canRunJobWithModel(ctx context.Context, h InterfaceWithModels, j workerStar
...
 
-	return h.CanSpawn(ctx, model, j.id, j.requirements)
+	return h.CanSpawn(ctx, sdk.WorkerStarterWorkerModel{ModelV1: model}, j.id, j.requirements)
 }
 
 // SendSpawnInfo sends a spawnInfo
-func SendSpawnInfo(ctx context.Context, h Interface, jobID int64, spawnMsg sdk.SpawnMsg) {
-	if h.CDSClient() == nil || jobID == 0 {
+func SendSpawnInfo(ctx context.Context, h Interface, jobID string, spawnMsg sdk.SpawnMsg) {
+	if h.CDSClient() == nil || jobID == "" {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply sdk/hatchery/hatchery.go:763]: fixed

File: engine/hatchery/vsphere/hatchery.go (hunk 1 above)
Comment (sguiheux) [reply engine/hatchery/vsphere/hatchery.go:152]: done

File: engine/hatchery/swarm/swarm.go (hunk 3 above)
Comment (sguiheux) [reply engine/hatchery/swarm/swarm.go:215]: done

File: engine/hatchery/local/worker_spawn.go (hunk 4 above)
Comment (sguiheux) [reply engine/hatchery/local/worker_spawn.go:31]: done

File: engine/hatchery/kubernetes/kubernetes.go (hunk 5 above)
Comment (sguiheux) [reply engine/hatchery/kubernetes/kubernetes.go:230]: done

File: engine/hatchery/kubernetes/kubernetes.go (hunk 6 above)
Comment (sguiheux) [reply engine/hatchery/kubernetes/kubernetes.go:225]: done

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18998.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Pull request merged**